src/gf_ring.h

```diff
@@ -148,12 +148,24 @@ class RingModN {
     std::vector<int> exponents;
     std::vector<T> all_primes_factors;
     std::vector<T> proper_divisors;
+    // random number engine with a seed as a random number
+    mutable std::uniform_int_distribution<uint64_t> distribution;
+    // random number engine for the __uint128_t case
+    mutable std::uniform_int_distribution<uint64_t> distribution_leading;
 };
 template <typename T>
 RingModN<T>::RingModN(T card)
+    : _card(card), root(0), distribution(
+                                1,
+                                (card > std::numeric_limits<uint64_t>::max())
```

Comment:
This ternary could be moved into a little helper to improve the readability:

```
[…]
  private:
    static constexpr uint64_t rand_upper_bound(T card)
    {
        return (card > std::numeric_limits<uint64_t>::max())
                   ? std::numeric_limits<uint64_t>::max()
                   : card - 1;
    }
};
template <typename T>
RingModN<T>::RingModN(T card)
    : _card(card), root(0), distribution(1, rand_upper_bound(card))
{
}

[…]
```

Reply:
yeah, thank you